static/js/message_list_view.js

```diff
@@ -1445,4 +1445,103 @@ export class MessageListView {
             message_container.status_message = false;
         }
     }
+
+    /* This function exist for two purposes:
+        * To track the current `sticky_header` which have some different properties
+          like date being always displayed.
+        * Set date on message header corresponding to the message next to the header. */
+    update_sticky_recipient_headers() {
+        const rows_length = this._rows.size;
+        if (!rows_length) {
+            /* No headers are present */
+            return;
+        }
+        /* Intentionally remove sticky headers class here to make calculations simpler. */
+        $(".sticky_header").removeClass("sticky_header");
+        /* visible_top is navbar top position + height for us. */
+        const visible_top = message_viewport.message_viewport_info().visible_top;
+        /* We need date to be properly visible on the header, so partially visible headers
+           who are about to be scrolled out of view are not acceptable. */
+        const header_about_to_be_partially_hidden = visible_top - 1;
+
+        function is_sticky(header) {
+            const header_props = header.getBoundingClientRect();
+            const margin_between_recipient_rows = 10;
+            const header_is_sticky_or_about_to_be_sticky =
+                visible_top + header_props.height + margin_between_recipient_rows;
+            if (header_props.top < header_about_to_be_partially_hidden) {
+                return -1;
+            } else if (header_props.top > header_is_sticky_or_about_to_be_sticky) {
+                return 1;
+            }
+            /* Headers between `header_about_to_be_partially_hidden` and `header_is_sticky_or_about_to_be_sticky`
+               are sticky. If two headers next to each other are completely visible
+               (message header at top has no visible content), we don't mind showing
+               date on any of them. Which header is chosen will depend on which
+               comes first when iterating on the headers. */
+            return 0;
+        }
```

Comment:
As a side note, I got curious about whether there's a standard way to check if a sticky header is stuck, and found this:

https://stackoverflow.com/questions/25308823/targeting-positionsticky-elements-that-are-currently-in-a-stuck-state

I have not investigated whether that technique would help us in any way, but figured it worth posting for your consideration.

Reply:
Yeah, I looked at it earlier, and it is not of any use to us.